src/effects/VST3/VST3EffectsModule.cpp

```diff
@@ -0,0 +1,320 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file VST3EffectsModule.cpp
+
+  @author Vitaly Sverchinsky
+
+  @brief Part of Audacity VST3 module
+
+**********************************************************************/
+
+
+#include "VST3EffectsModule.h"
+
+#include <stdexcept>
+
+#include <wx/stdpaths.h>
+#include <wx/dir.h>
+#include <wx/log.h>
+#include <public.sdk/source/vst/hosting/module.h>
+#include <public.sdk/source/vst/hosting/plugprovider.h>
+#include <pluginterfaces/vst/ivstaudioprocessor.h>
+
+
+#include "../../ModuleManager.h"
+
+#include "wxArrayStringEx.h"
+#include "PluginInterface.h"
+#include "ModuleInterface.h"
+#include "VST3Effect.h"
+#include "VST3Utils.h"
+#include "widgets/AButton.h"
+
+
+DECLARE_MODULE_ENTRY(AudacityModule)
+{
+   // Create our effects module and register
+   // Trust the module manager not to leak this
+   return safenew VST3EffectsModule();
+}
+
+DECLARE_BUILTIN_MODULE(VST3Builtin);
+
+//Helper class used to find vst3 effects during folder traversal
+class VST3PluginTraverser final : public wxDirTraverser
+{
+   std::function<void(const wxString&)> mOnPluginFound;
+public:
+   VST3PluginTraverser(std::function<void(const wxString&)> onPluginFound)
+      : mOnPluginFound(std::move(onPluginFound))
+   {
+   }
+
+   wxDirTraverseResult OnFile(const wxString& filename) override
+   {
+      if(filename.Matches("*.vst3"))
+         mOnPluginFound(filename);
+      return wxDIR_CONTINUE;
+   }
+   wxDirTraverseResult OnDir(const wxString& dirname) override
+   {
+      if(dirname.Matches("*.vst3"))
+      {
+         mOnPluginFound(dirname);
+         return wxDIR_IGNORE;//do not look inside...
+      }
+      return wxDIR_CONTINUE;
+   }
+};
+
+const TranslatableString VST3EffectsModule::PluginType = XO("VST3");
+
+std::shared_ptr<VST3::Hosting::Module> VST3EffectsModule::GetModule(const wxString& path)
+{
+   const auto it = mModules.find(path);
+   if(it != mModules.end())
+   {
+      if(auto lock = it->second.lock())
+         return lock;
+   }
+
+   std::string moduleCreateError;
+   auto module = VST3::Hosting::Module::create(path.ToStdString(), moduleCreateError);
```

Comment:
Let's comment that this line is the (sole?) entry into platform-specific code in one of three conditionally compiled files.